File: scripts/vivid-publish-blog.mjs
#!/usr/bin/env node
/**
 * Vivid Health — publish blog + articles to Shopify.
 *
 * Reads vivid/data/journal.json and:
 *   1. Creates (or reuses) a blog with handle "journal"
 *   2. Upserts each article (idempotent — looks up by handle, updates or creates)
 *   3. Tags articles with their category for filtering
 *   4. Uploads the hero image if it's a relative path under vivid/assets
 *
 * Requires Shopify Admin API scopes: write_content (currently NOT granted
 * on the Vivid Claude Review app — extend or use a separate write-scoped app).
 *
 * Env:
 *   SHOPIFY_STORE           e.g. hgywg0-w7.myshopify.com
 *   SHOPIFY_API_VERSION     e.g. 2026-04
 *   SHOPIFY_CLIENT_ID       client credentials
 *   SHOPIFY_CLIENT_SECRET
 *   DRY_RUN=1               print actions without executing
 */

import { readFile } from "node:fs/promises";
import { resolve, dirname } from "node:path";
import { fileURLToPath } from "node:url";

const STORE = required("SHOPIFY_STORE");
const API = required("SHOPIFY_API_VERSION");
const CLIENT_ID = required("SHOPIFY_CLIENT_ID");
const CLIENT_SECRET = required("SHOPIFY_CLIENT_SECRET");
const DRY_RUN = process.env.DRY_RUN === "1";
const REPO_ROOT = resolve(dirname(fileURLToPath(import.meta.url)), "..");
const JOURNAL_JSON = resolve(REPO_ROOT, "vivid", "data", "journal.json");
const BLOG_HANDLE = "journal";
const BLOG_TITLE = "Journal";
const COMMIT_POLICY_NOTE = "All articles authored by Vivid Lab Team. Re-publishing this script is idempotent.";

function required(k) {
  const v = process.env[k];
  if (!v) { console.error("Missing env: " + k); process.exit(1); }
  return v;
}

async function tokenExchange() {
  const r = await fetch(`https://${STORE}/admin/oauth/access_token`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ client_id: CLIENT_ID, client_secret: CLIENT_SECRET, grant_type: "client_credentials" }),
  });
  const j = await r.json();
  if (!j.access_token) throw new Error("token exchange failed: " + JSON.stringify(j));
  return j.access_token;
}

function api(token) {
  return async function (query, variables = {}) {
    const r = await fetch(`https://${STORE}/admin/api/${API}/graphql.json`, {
      method: "POST",
      headers: { "X-Shopify-Access-Token": token, "Content-Type": "application/json" },
      body: JSON.stringify({ query, variables }),
    });
    const j = await r.json();
    if (j.errors) throw new Error("GraphQL: " + JSON.stringify(j.errors));
    return j.data;
  };
}

async function ensureBlog(gql) {
  const found = await gql(`{ blogs(first: 50) { edges { node { id handle title } } } }`);
  const existing = found.blogs.edges.find((e) => e.node.handle === BLOG_HANDLE);
  if (existing) { console.log(`  · blog "${BLOG_HANDLE}" exists → ${existing.node.id}`); return existing.node.id; }
  if (DRY_RUN) { console.log(`  · DRY_RUN: would create blog "${BLOG_HANDLE}"`); return "gid://shopify/Blog/DRY"; }
  const res = await gql(`mutation($input: BlogCreateInput!) { blogCreate(blog: $input) { blog { id handle title } userErrors { field message } } }`, {
    input: { title: BLOG_TITLE, handle: BLOG_HANDLE, commentPolicy: "MODERATED" },
  });
  const errs = res.blogCreate?.userErrors || [];
  if (errs.length) throw new Error("blogCreate: " + JSON.stringify(errs));
  console.log(`  · created blog "${BLOG_HANDLE}" → ${res.blogCreate.blog.id}`);
  return res.blogCreate.blog.id;
}

async function listArticles(gql, blogId) {
  const all = new Map();
  let cursor = null;
  for (;;) {
    const r = await gql(`query($id: ID!, $after: String) {
      blog(id: $id) {
        articles(first: 50, after: $after) {
          edges { node { id handle title isPublished tags } }
          pageInfo { hasNextPage endCursor }
        }
      }
    }`, { id: blogId, after: cursor });
    for (const e of r.blog.articles.edges) all.set(e.node.handle, e.node);
    if (!r.blog.articles.pageInfo.hasNextPage) break;
    cursor = r.blog.articles.pageInfo.endCursor;
  }
  return all;
}

function authorBody(article) {
  const head = `<p class="article-sub">${article.sub}</p>`;
  const meta = `<p class="article-meta"><strong>${article.cat}</strong> · ${article.read} · ${article.date}</p>`;
  let body = article.body_html;
  // Wrap in <article class="rte"> for theme styling.
  const related = (article.related || []).length
    ? `<aside class="article-pairings"><h3>Pairs well with</h3><ul>${article.related.map(h => `<li><a href="/products/${h}">${h}</a></li>`).join("")}</ul></aside>`
    : "";
  return `${meta}\n${head}\n${body}\n${related}`;
}

async function upsertArticle(gql, blogId, article, existing) {
  const body = authorBody(article);
  const tags = [article.cat || "", "vivid-journal", "category:" + (article.cat || "").toLowerCase().replace(/\s+/g, "-")].filter(Boolean);
  if (existing) {
    if (DRY_RUN) { console.log(`  · DRY_RUN: would UPDATE ${article.handle}`); return; }
    const r = await gql(`mutation($id: ID!, $article: ArticleUpdateInput!) {
      articleUpdate(id: $id, article: $article) {
        article { id handle title }
        userErrors { field message }
      }
    }`, {
      id: existing.id,
      article: { title: article.title, body, summary: article.sub, tags, author: { name: article.author || "Vivid Lab Team" } },
    });
    const errs = r.articleUpdate?.userErrors || [];
    if (errs.length) throw new Error(`articleUpdate ${article.handle}: ` + JSON.stringify(errs));
    console.log(`  · updated  ${article.handle}`);
  } else {
    if (DRY_RUN) { console.log(`  · DRY_RUN: would CREATE ${article.handle}`); return; }
    const r = await gql(`mutation($article: ArticleCreateInput!) {
      articleCreate(article: $article) {
        article { id handle title }
        userErrors { field message }
      }
    }`, {
      article: {
        blogId,
        title: article.title,
        handle: article.handle,
        body,
        summary: article.sub,
        tags,
        author: { name: article.author || "Vivid Lab Team" },
        isPublished: true,
      },
    });
    const errs = r.articleCreate?.userErrors || [];
    if (errs.length) throw new Error(`articleCreate ${article.handle}: ` + JSON.stringify(errs));
    console.log(`  · created  ${article.handle}`);
  }
}

async function main() {
  console.log(`Vivid blog publish ${DRY_RUN ? "(DRY_RUN)" : ""}\n`);
  const articles = JSON.parse(await readFile(JOURNAL_JSON, "utf8"));
  console.log(`Source: ${JOURNAL_JSON} — ${articles.length} articles\n`);

  const token = await tokenExchange();
  const gql = api(token);

  console.log("Ensure blog exists:");
  const blogId = await ensureBlog(gql);

  console.log("\nList existing articles:");
  const existing = DRY_RUN ? new Map() : await listArticles(gql, blogId);
  console.log(`  ${existing.size} existing in "${BLOG_HANDLE}"`);

  console.log("\nUpsert articles:");
  let ok = 0, fail = 0;
  for (const a of articles) {
    try { await upsertArticle(gql, blogId, a, existing.get(a.handle)); ok++; }
    catch (e) { console.error(`  ✗ ${a.handle}: ${e.message}`); fail++; }
  }

  console.log(`\nDone. ok=${ok}  fail=${fail}`);
  console.log(`\n${COMMIT_POLICY_NOTE}`);
  process.exit(fail > 0 ? 1 : 0);
}

main().catch((e) => { console.error(e); process.exit(2); });

File: scripts/vivid_publish_blog.py
#!/usr/bin/env python3
"""Publish one queued markdown article to the Vivid Health store's Journal.

Usage: python3 scripts/vivid_publish_blog.py content/vivid-queue/2026-07-foo.md

Auth: VIVID_CLIENT_ID + VIVID_CLIENT_SECRET (client credentials) or
VIVID_ADMIN_TOKEN, against VIVID_STORE (default hgywg0-w7).
Finds the blog with handle 'journal' (falls back to the first blog),
creates the article published, prints the URL. The workflow moves the
file to content/vivid-published/ afterwards.

Frontmatter: title, slug, excerpt, tags, products (list of handle/name/price).
Markdown subset: #/##/### headings, **bold**, *italic*, [links](), lists,
tables, paragraphs — same conventions as scripts/publish_blog.py.
"""
import json
import os
import re
import sys
import urllib.error
import urllib.parse
import urllib.request

API = "2025-01"
STORE = (os.environ.get("VIVID_STORE") or "hgywg0-w7").replace(".myshopify.com", "").strip()
BASE = f"https://{STORE}.myshopify.com"


def get_token():
    t = (os.environ.get("VIVID_ADMIN_TOKEN") or "").strip()
    if t:
        return t
    cid = (os.environ.get("VIVID_CLIENT_ID") or "").strip()
    csec = (os.environ.get("VIVID_CLIENT_SECRET") or "").strip()
    if not (cid and csec):
        sys.exit("ERROR: VIVID_ADMIN_TOKEN or VIVID_CLIENT_ID+SECRET required")
    body = urllib.parse.urlencode({
        "grant_type": "client_credentials", "client_id": cid, "client_secret": csec,
    }).encode()
    req = urllib.request.Request(
        f"{BASE}/admin/oauth/access_token", data=body,
        headers={"Content-Type": "application/x-www-form-urlencoded"}, method="POST")
    with urllib.request.urlopen(req, timeout=30) as r:
        return json.loads(r.read())["access_token"]


TOKEN = None


def api(method, path, payload=None):
    url = BASE + path
    data = json.dumps(payload).encode() if payload is not None else None
    req = urllib.request.Request(url, data=data, method=method, headers={
        "X-Shopify-Access-Token": TOKEN, "Content-Type": "application/json",
        "Accept": "application/json"})
    try:
        with urllib.request.urlopen(req, timeout=60) as r:
            return r.status, json.loads(r.read() or b"{}")
    except urllib.error.HTTPError as e:
        return e.code, {"error": e.read().decode()[:400]}


# ---------- frontmatter + markdown (conventions from publish_blog.py) ----------

def parse(md):
    m = re.match(r"^---\n(.*?)\n---\n(.*)$", md, re.S)
    if not m:
        sys.exit("ERROR: missing frontmatter")
    raw, body = m.group(1), m.group(2)
    fm, cur_list, cur_key = {}, None, None
    for line in raw.split("\n"):
        if re.match(r"^\s*-\s+handle:", line) and cur_key:
            cur_list.append({"handle": line.split("handle:", 1)[1].strip()})
        elif re.match(r"^\s+(name|price):", line) and cur_list:
            k, v = line.strip().split(":", 1)
            cur_list[-1][k] = v.strip()
        elif re.match(r"^[A-Za-z_]+:", line):
            k, v = line.split(":", 1)
            k, v = k.strip(), v.strip().strip('"').strip("'")
            if v == "":
                cur_key, cur_list = k, []
                fm[k] = cur_list
            else:
                fm[k] = v
                cur_key, cur_list = None, None
    return fm, body


def _inline(t):
    t = re.sub(r"\*\*(.+?)\*\*", r"<strong>\1</strong>", t)
    t = re.sub(r"(?<!\*)\*(?!\*)(.+?)(?<!\*)\*(?!\*)", r"<em>\1</em>", t)
    t = re.sub(r"\[([^\]]+)\]\(([^)]+)\)", r'<a href="\2">\1</a>', t)
    return t


def md_to_html(body):
    out, para, in_list = [], [], False

    def flush_para():
        nonlocal para
        if para:
            out.append("<p>" + _inline(" ".join(para)) + "</p>")
            para = []

    def close_list():
        nonlocal in_list
        if in_list:
            out.append("</ul>")
            in_list = False

    lines = body.split("\n")
    i = 0
    while i < len(lines):
        line = lines[i].rstrip()
        if re.match(r"^\|.+\|$", line) and i + 1 < len(lines) and re.match(r"^\|[\s\-:|]+\|$", lines[i + 1].rstrip()):
            flush_para(); close_list()
            headers = [c.strip() for c in line.strip("|").split("|")]
            out.append('<table><thead><tr>' + "".join(f"<th>{_inline(h)}</th>" for h in headers) + "</tr></thead><tbody>")
            i += 2
            while i < len(lines) and re.match(r"^\|.+\|$", lines[i].rstrip()):
                cells = [c.strip() for c in lines[i].rstrip().strip("|").split("|")]
                out.append("<tr>" + "".join(f"<td>{_inline(c)}</td>" for c in cells) + "</tr>")
                i += 1
            out.append("</tbody></table>")
            continue
        if line.startswith("### "):
            flush_para(); close_list(); out.append(f"<h3>{_inline(line[4:])}</h3>")
        elif line.startswith("## "):
            flush_para(); close_list(); out.append(f"<h2>{_inline(line[3:])}</h2>")
        elif line.startswith("# "):
            flush_para(); close_list(); out.append(f"<h2>{_inline(line[2:])}</h2>")
        elif re.match(r"^[-*] ", line):
            flush_para()
            if not in_list:
                out.append("<ul>")
                in_list = True
            out.append(f"<li>{_inline(line[2:])}</li>")
        elif line.strip() == "":
            flush_para(); close_list()
        else:
            para.append(line.strip())
        i += 1
    flush_para(); close_list()
    return "\n".join(out)


def render_products(products):
    if not products:
        return ""
    cards = []
    for p in products:
        h = p.get("handle", "")
        cards.append(
            f'<div style="border:1px solid #E4DCC9;border-radius:12px;padding:16px 18px;margin:10px 0;background:#FBF7EE">'
            f'<strong>{p.get("name", h)}</strong>'
            f'<span style="float:right;font-weight:600">{p.get("price", "")}</span><br>'
            f'<a href="/products/{h}">View product →</a></div>')
    return ('<h3>From the range</h3>' + "".join(cards) +
            '<p><em>All prices include VAT. Not sure what you need? '
            '<a href="/pages/quiz">Take the 2-minute quiz</a> or '
            '<a href="/pages/consultation">book a consultation</a>.</em></p>')


def verify_products(products):
    ok = []
    for p in products or []:
        h = p.get("handle", "")
        try:
            with urllib.request.urlopen(f"{BASE}/products/{h}.js", timeout=20) as r:
                data = json.loads(r.read())
                if data.get("available") is False:
                    print(f"  ⚠ {h} exists but SOLD OUT — dropping from card list")
                    continue
                ok.append(p)
        except Exception:
            print(f"  ⚠ product {h} not found on storefront — dropping")
    return ok


def main():
    global TOKEN
    if len(sys.argv) < 2:
        sys.exit("usage: vivid_publish_blog.py <content/vivid-queue/file.md>")
    fm, body = parse(open(sys.argv[1]).read())
    for k in ("title", "slug", "excerpt"):
        if not fm.get(k):
            sys.exit(f"ERROR: frontmatter missing {k}")

    TOKEN = get_token()
    print("✓ Token acquired")

    code, blogs = api("GET", f"/admin/api/{API}/blogs.json")
    if code != 200:
        sys.exit(f"ERROR: blogs.json HTTP {code}: {blogs}")
    blog = next((b for b in blogs["blogs"] if b["handle"] == "journal"),
                (blogs["blogs"] or [None])[0])
    if not blog:
        sys.exit("ERROR: store has no blog")
    print(f"✓ Blog: {blog['handle']} ({blog['id']})")

    products = verify_products(fm.get("products"))
    html = md_to_html(body) + render_products(products)
    html += '<p>— The Vivid formulation desk</p>'

    payload = {"article": {
        "title": fm["title"],
        "handle": fm["slug"],
        "body_html": html,
        "tags": fm.get("tags", ""),
        "summary_html": f"<p>{fm['excerpt']}</p>",
        "published": True,
    }}
    code, res = api("POST", f"/admin/api/{API}/blogs/{blog['id']}/articles.json", payload)
    if code not in (200, 201):
        sys.exit(f"ERROR: article create HTTP {code}: {res}")
    art = res["article"]
    url = f"{BASE}/blogs/{blog['handle']}/{art['handle']}"
    print(f"✓ Published: {url}")
    with open("vivid-blog-last-published.txt", "w") as f:
        f.write(url + "\n")


if __name__ == "__main__":
    main()
